addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java

```diff
@@ -0,0 +1,176 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.evohome.internal.api;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.evohome.internal.api.models.v2.response.Authentication;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Provides access to (an optionally OAUTH based) API. Makes sure that all the necessary headers are set.
+ *
+ * @author Jasper van Zuijlen - Initial contribution
+ *
+ */
+public class ApiAccess {
+    private final Logger logger = LoggerFactory.getLogger(ApiAccess.class);
+    private final HttpClient httpClient;
+
+    private Authentication authenticationData;
+    private String applicationId;
+
+    public ApiAccess(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Sets the authentication details on the type
+     *
+     * @param authentication The authentication details to apply
+     */
+    public void setAuthentication(Authentication authentication) {
+        authenticationData = authentication;
+    }
+
+    /**
+     * Gets the current authentication details of the type
+     *
+     * @return The current authentication details
+     */
+    public Authentication getAuthentication() {
+        return authenticationData;
+    }
+
+    /**
+     * Sets the application id on the type
+     *
+     * @param applicationId The application id to apply
+     */
+    public void setApplicationId(String applicationId) {
+        applicationId = applicationId;
+    }
+
+    /**
+     * Issues an HTTP request on the API's URL. Makes sure that the request is correctly formatted.
+     *
+     * @param method The HTTP method to use (POST, GET, ...)
+     * @param url The URL to query
+     * @param headers The optional additional headers to apply, can be null
+     * @param requestData The optional request data to use, can be null
+     * @param contentType The content type to use with the request data. Required when using requestData
+     * @param out An instance of the return type, ignored when value is null
+     * @return The result of the request or null
+     */
+    @SuppressWarnings("unchecked")
+    public <TOut> TOut doRequest(HttpMethod method, String url, Map<String, String> headers, String requestData,
+            String contentType, TOut out) {
```

Comment:
I changed the signature to `public <TOut> TOut doRequest(HttpMethod method, String url, Map<String, String> headers, String requestData, String contentType, Class<TOut> outClass, TOut out)`

This mitigates the need for unchecked since `retVal = new Gson().fromJson(reply, outClass);` now returns the actual TOut class and I don't have to hard-cast anymore.